src/goto-programs/goto_cfg.cpp

```diff
@@ -165,3 +194,646 @@ void goto_cfg::dump_graph() const
     file << "}\n";
   }
 }
+
+template <class F>
+void goto_cfg::basic_block::foreach_inst(F f)
+{
+  for (goto_programt::instructionst::iterator start = begin; start != end;
+       start++)
+    f(*start);
+}
+
+template <class F>
+void goto_cfg::basic_block::foreach_bb(F f)
+{
+  goto_cfg::foreach_bb(this, f);
+}
+
+template <class F>
+void goto_cfg::foreach_bb(
+  const std::shared_ptr<goto_cfg::basic_block> &start,
+  F foo)
+{
+  std::unordered_set<std::shared_ptr<goto_cfg::basic_block>> visited;
+  std::vector<std::shared_ptr<goto_cfg::basic_block>> to_visit({start});
+
+  while (!to_visit.empty())
+  {
+    const auto &item = to_visit.back();
+    visited.insert(item);
+    to_visit.pop_back();
+    foo(item);
+    for (const auto &next : item->successors)
+      if (!visited.count(next))
+        to_visit.push_back(next);
+  }
+}
+
+void Dominator::compute_dominators()
+{
+  // TODO: This algorithm is quadratic, there is a more efficient
+  // version by Lengauer-Tarjan.
+
+  // Computes dominator of a node
+  std::unordered_map<CFGNode, std::unordered_set<CFGNode>> dt;
+
+  // 1. Sets all nodes dominators to TOP
+
+  std::unordered_set<CFGNode> top;
+  goto_cfg::foreach_bb(start, [&top](const CFGNode &n) { top.insert(n); });
+
+  goto_cfg::foreach_bb(start, [&dt, &top](const CFGNode &n) { dt[n] = top; });
+
+  /*
+    Dom(n) = {n}, if n = start
+             {n} `union` (intersec_pred Dom(p))
+   */
+  dt[start] = std::unordered_set<CFGNode>({start});
+
+  std::vector<std::shared_ptr<goto_cfg::basic_block>> worklist{start};
+  for (const auto &suc : start->successors)
+    worklist.push_back(suc);
+
+  while (!worklist.empty())
+  {
+    const auto &dominator_node = worklist.back();
+    worklist.pop_back();
+
+    if (dominator_node == start)
+      continue;
+
+    // Get the intersection of all the predecessors
+    std::unordered_set<CFGNode> intersection = top;
+    for (const auto &pred : dominator_node->predecessors)
+    {
+      assert(dt.count(pred));
+      for (const auto &item : top)
+        if (!dt[pred].count(item))
+          intersection.erase(item);
+    }
+
+    // Union of node with its predecessor intersection
+    std::unordered_set<CFGNode> result({dominator_node});
+    for (const auto &item : intersection)
+      result.insert(item);
+
+    // Fix-point?
+    if (dt[dominator_node] != result)
+    {
+      dt[dominator_node] = result;
+      for (const auto &suc : dominator_node->successors)
+        worklist.push_back(suc);
+    }
+  }
+
+  _dominators = dt;
+}
+
+void Dominator::dump_dominators() const
+{
+  for (const auto &[node, edges] : _dominators)
+  {
+    log_status("CFGNode");
+    node->begin->dump();
+
+    log_status("Dominated");
+    for (const auto &edge : edges)
+      edge->begin->dump();
+  }
+}
+
+CFGNode Dominator::idom(const CFGNode &n) const
```

Comment:
Add some documentation here.

Reply:
The header contains the doc for this method